File: day10/Main.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.text.NumberFormat;
import java.text.ParseException;
import java.util.*;

class Position {
    public Position(int row, int col) {
        this.row = row;
        this.col = col;
    }

    int row;
    int col;
}

public class Main {

    public static int stoi(String s, int beginIndex) throws ParseException {
        NumberFormat nf = NumberFormat.getInstance();
        return nf.parse((s.substring(beginIndex)).strip()).intValue();
    }

    public static void part1(BufferedReader br) throws Exception {
        List<char[]> gridTmp = new ArrayList<>();
        String line;
        while ((line = br.readLine()) != null) {
            gridTmp.add(line.toCharArray());
        }
        char[][] grid = gridTmp.toArray(char[][]::new);

        Map<Character, int[][]> symbolToMoves = new HashMap<>();
        symbolToMoves.put('|', new int[][]{{-1, 0}, {1, 0}});
        symbolToMoves.put('-', new int[][]{{0, -1}, {0, 1}});
        symbolToMoves.put('L', new int[][]{{-1, 0}, {0, 1}});
        symbolToMoves.put('J', new int[][]{{-1, 0}, {0, -1}});
        symbolToMoves.put('7', new int[][]{{0, -1}, {1, 0}});
        symbolToMoves.put('F', new int[][]{{1, 0}, {0, 1}});

        // get start position
        Position start = new Position(-1, -1);
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid.length; j++) {
                if (grid[i][j] == 'S') {
                    start.row = i;
                    start.col = j;
                    break;
                }
            }
        }

        grid[start.row][start.col] = 'F'; // fixme!

        int loopSize = 0;
        boolean[][] pipes = new boolean[grid.length][grid[0].length];
        Position curr = start;
        Position prev = new Position(-1, -1);

        while (!pipes[curr.row][curr.col]) {
            loopSize++;
            pipes[curr.row][curr.col] = true;

            int[][] moves = symbolToMoves.get(grid[curr.row][curr.col]);
            for (int[] move : moves) {
                int row = curr.row + move[0];
                int col = curr.col + move[1];

                if (row != prev.row || col != prev.col) {
                    prev = curr;
                    curr = new Position(row, col);
                    break;
                }

            }
        }

        System.out.println("loop size: " + loopSize);
        System.out.println("farthest from start: " + loopSize / 2);
    }


    public static void part2(BufferedReader br) throws Exception {
        List<char[]> gridTmp = new ArrayList<>();
        String line;
        while ((line = br.readLine()) != null) {
            gridTmp.add(line.toCharArray());
        }
        char[][] grid = gridTmp.toArray(char[][]::new);

        Map<Character, int[][]> symbolToMoves = new HashMap<>();
        symbolToMoves.put('|', new int[][]{{-1, 0}, {1, 0}});
        symbolToMoves.put('-', new int[][]{{0, -1}, {0, 1}});
        symbolToMoves.put('L', new int[][]{{-1, 0}, {0, 1}});
        symbolToMoves.put('J', new int[][]{{-1, 0}, {0, -1}});
        symbolToMoves.put('7', new int[][]{{0, -1}, {1, 0}});
        symbolToMoves.put('F', new int[][]{{1, 0}, {0, 1}});

        // get start position
        Position start = new Position(-1, -1);
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] == 'S') {
                    start.row = i;
                    start.col = j;
                    break;
                }
            }
        }

        grid[start.row][start.col] = 'F'; // fixme: starting symbol

        int loopSize = 0;
        boolean[][] inLoop = new boolean[grid.length][grid[0].length];
        Position curr = start;
        Position prev = new Position(-1, -1);

        while (!inLoop[curr.row][curr.col]) {
            loopSize++;
            inLoop[curr.row][curr.col] = true;

            int[][] moves = symbolToMoves.get(grid[curr.row][curr.col]);
            for (int[] move : moves) {
                int row = curr.row + move[0];
                int col = curr.col + move[1];

                if (row != prev.row || col != prev.col) {
                    prev = curr;
                    curr = new Position(row, col);
                    break;
                }

            }
        }

        // part one answer
        System.out.println("loop size: " + loopSize);
        System.out.println("farthest from start: " + loopSize / 2);

        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[i].length; j++) {
                if (!inLoop[i][j]) {
                    grid[i][j] = '.';
                }
                if (grid[i][j] == '7' || grid[i][j] == 'F') {
                    grid[i][j] = '|';
                }
            }
        }

        int enclosed = 0;
        for (char[] row : grid) {
            int pipes = 0;
            for (char symbol : row) {
                if (symbol == '|') {
                    pipes++;
                }
                if (pipes % 2 == 1 && symbol == '.') {
                    enclosed++;
                }
            }
        }

        System.out.println("enclosed: " + enclosed);
    }


    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new FileReader("jaoc23/src/input"));
//        part1(br);
        part2(br);

    }
}

File: day10/main.cpp
#include <iostream>
#include <string>
#include <vector>
#include <fstream>
#include <deque>
#include <numeric>
#include "utils_aoc.h"

using namespace std;

//
//int get_size_loop(const pair<int, int> &start, const vector<string> &grid,
//                  unordered_map<char, vector<vector<int>>> &sign_to_moves, int half = -1) {
//    int loop_size = 0;
//    int start_visits = 0;
//    pair<int, int> curr = start;
//    pair<int, int> prev = {-1, -1};
//    while (loop_size < 1000) {
//        if (curr.first == start.first && curr.second == start.second) {
//            start_visits++;
//            if (start_visits > 1) {
//                cout << "starts" << start_visits << endl;
//                break;
//            }
//        }
//        loop_size++;
////        if (loop_size == half) {
////            break;
////        }
//
//        auto moves = sign_to_moves[grid[curr.first][curr.second]];
//        for (const auto &move: moves) {
//            int row = curr.first + move[0];
//            int col = curr.second + move[1];
//            if (row < 0 || col < 0 || row >= grid.size() || col >= grid[0].size()) {
//                cout << "err: " << row << " " << col << endl;
//                return 0;
//            }
//            if (pair<int, int>{row, col} != prev) {
//                prev = curr;
//                curr = {row, col};
//                break;
//            }
//        }
//        cout << "curr: " << curr.first << " " << curr.second << " " << grid[curr.first][curr.second] << endl;
//
//    }
//}

void part1(ifstream &input) {
    string line;
    vector<string> grid;
    while (getline(input, line)) {
        grid.push_back(line);
    }

    pair<int, int> start;
    for (int i = 0; i < grid.size(); i++) {
        for (int j = 0; j < grid.size(); j++) {
            if (grid[i][j] == 'S') {
                start = {i, j};
                break;
            }
        }
    }

    unordered_map<char, vector<vector<int>>> sign_to_moves{
            {'|', {{-1, 0},  {1, 0}}},
            {'-', {{0,  -1}, {0, 1}}},
            {'L', {{-1, 0},  {0, 1}}},
            {'J', {{-1, 0},  {0, -1}}},
            {'7', {{0,  -1}, {1, 0}}},
            {'F', {{1,  0},  {0, 1}}},
            {'.', {}}
    };

    long long ans = 0;


    char start_sign = '|'; // fixme
    grid[start.first][start.second] = start_sign;
    cout << start.first << " " << start.second << " " << start_sign << endl;

    // get size of loop
//    int loop_size = get_size_loop(start, grid, sign_to_moves);

    int loop_size = 0;
    int start_visits = 0;
    pair<int, int> curr = start;
    pair<int, int> prev = {-1, -1};
    while (true) {
        if (curr.first == start.first && curr.second == start.second) {
            start_visits++;
            if (start_visits > 1) {
                cout << "starts" << start_visits << endl;
                break;
            }
        }
        loop_size++;
//        if (loop_size == half) {
//            break;
//        }

        auto moves = sign_to_moves[grid[curr.first][curr.second]];
        for (const auto &move: moves) {
            int row = curr.first + move[0];
            int col = curr.second + move[1];
            if (row < 0 || col < 0 || row >= grid.size() || col >= grid[0].size()) {
                cout << "err: " << row << " " << col << endl;
                return;
            }
            if (pair<int, int>{row, col} != prev) {
                prev = curr;
                curr = {row, col};
                break;
            }
        }
        cout << "curr: " << curr.first << " " << curr.second << " " << grid[curr.first][curr.second] << endl;

    }

    cout << "loop_size: " << loop_size << endl;

    int half = loop_size / 2;
    if (loop_size % 2 == 1) {
        half++;
    }
    cout << "half: " << half << endl;
}

void part2(ifstream &input) {
    string line;
    vector<string> grid;
    while (getline(input, line)) {
        grid.push_back(line);
    }

    pair<int, int> start = {-1, -1};
    for (int i = 0; i < grid.size(); i++) {
        for (int j = 0; j < grid[i].size(); j++) {
            if (grid[i][j] == 'S') {
                start = {i, j};
                break;
            }
        }
    }
    if (start.first < 0) {
        cout << " err: " << start.first << " " << start.second << endl;
        return;
    }

    unordered_map<char, vector<vector<int>>> sign_to_moves{
            {'|', {{-1, 0},  {1, 0}}},
            {'-', {{0,  -1}, {0, 1}}},
            {'L', {{-1, 0},  {0, 1}}},
            {'J', {{-1, 0},  {0, -1}}},
            {'7', {{0,  -1}, {1, 0}}},
            {'F', {{1,  0},  {0, 1}}},
            {'.', {{-1, 0},  {1, 0}, {0, -1}, {0, 1}}}
    };

    long long ans = 0;


    char start_sign = '|'; // fixme
    grid[start.first][start.second] = start_sign;
    cout << start.first << " " << start.second << " " << start_sign << endl;

    // mark pipe
    vector<vector<bool>> pipes(grid.size(), vector<bool>(grid[0].size()));
    int loop_size = 0;
    int start_visits = 0;
    pair<int, int> curr = start;
    pair<int, int> prev = {-1, -1};
    while (true) {
        if (curr.first == start.first && curr.second == start.second) {
            start_visits++;
            if (start_visits > 1) {
                cout << "starts" << start_visits << endl;
                break;
            }
        }
        loop_size++;

        auto moves = sign_to_moves[grid[curr.first][curr.second]];
        for (const auto &move: moves) {
            int row = curr.first + move[0];
            int col = curr.second + move[1];
            if (row < 0 || col < 0 || row >= grid.size() || col >= grid[0].size()) {
                cout << "err: " << row << " " << col << endl;
                return;
            }
            if (pair<int, int>{row, col} != prev) {
                prev = curr;
                curr = {row, col};
                pipes[row][col] = true;
                break;
            }
        }
//        cout << "curr: " << curr.first << " " << curr.second << " " << grid[curr.first][curr.second] << endl;

    }
    cout << "loop_size: " << loop_size << endl;


    for (int i = 0; i < grid.size(); i++) {
        for (int j = 0; j < grid[0].size(); j++) {
            if (!pipes[i][j]) {
                grid[i][j] = '.';
            }
            if (grid[i][j] == '7' || grid[i][j] == 'F') {
//            if (grid[i][j] != '.' && grid[i][j] != '|' && grid[i][j] != '-') {
                grid[i][j] = '|';
            }
        }
    }

    int counter = 0;
    for (int i = 0; i < grid.size(); i++) {
        int pipe_count = 0;
        for (int j = 0; j < grid[0].size(); j++) {
            if (grid[i][j] == '|') {
                pipe_count++;
            } else if (grid[i][j] == '.') {
                if (pipe_count % 2 == 1) {
                    cout << i << " " << j << endl;
                    counter++;
                }
            }
        }
    }


    int grid_total_size = grid.size() * grid[0].size();
    cout << "total: " << grid_total_size << endl;
    cout << "counter: " << counter << endl;
//    cout << "enclosed: " << grid_total_size - counter << endl;
}

int main() {
    ifstream input("/home/leartsura/Documents/projects/aoc23/input");
    if (input.fail()) {
        cout << "cant open input file" << endl;
        exit(0);
    }
//    part1(input);
    part2(input);

    input.close();
}
